lib/motsu/src/context/mod.rs

```diff
@@ -0,0 +1,188 @@
+    /// test thread.
+    pub fn reset_storage(self) {
+        EVM.remove(&self.thread_name);
+    }
+
+    /// Gets the code hash of the account at the given address.
+    pub fn account_codehash(self) -> [u8; 66] {
+        let context = EVM.entry(self.thread_name).or_default();
+        context.environment.account_codehash()
+    }
+
+    /// Gets a bounded estimate of the Unix timestamp at which the Sequencer
+    /// sequenced the transaction.
+    pub fn block_timestamp(self) -> u64 {
+        let context = EVM.entry(self.thread_name).or_default();
+        context.environment.block_timestamp()
+    }
+
+    /// Gets the chain ID of the current chain.
+    pub fn chain_id(self) -> u64 {
+        let context = EVM.entry(self.thread_name).or_default();
+        context.environment.chain_id()
+    }
+
+    /// Gets the address of the current program.
+    pub fn contract_address(self) -> [u8; 42] {
+        let context = EVM.entry(self.thread_name).or_default();
+        context.environment.contract_address()
+    }
+
+    /// Emits an EVM log with the given number of topics and data, the first
+    /// bytes of which should be the 32-byte-aligned topic data.
+    ///
+    /// Data contains `topics` amount of topics and then encoded event in `data`
+    /// buffer.
+    pub(crate) unsafe fn emit_log(
+        self,
+        data: *const u8,
+        len: usize,
+        topics: usize,
+    ) {
+        // https://github.com/OffchainLabs/stylus-sdk-rs/blob/v0.6.0/stylus-sdk/src/evm.rs#L38-L52
+        let buffer = read_bytes(data, len);
+        let encoded_event: Vec<u8> =
+            buffer.clone().into_iter().skip(topics * WORD_BYTES).collect();
+        panic!("Data: {:x?}, len: {:x?}, topics: {:x?}, Log: {:x?}, encoded_event: {:x?}", data, len, topics, buffer, encoded_event);
+        let mut context = EVM.entry(self.thread_name).or_default();
+        context.environment.store_event(&encoded_event);
+    }
+
+    /// Gets the address of the account that called the program.
+    pub fn msg_sender(self) -> [u8; 42] {
+        let context = EVM.entry(self.thread_name).or_default();
+        context.environment.msg_sender()
+    }
+
+    /// Removes all events for a test case.
+    pub fn clear_events(self) {
+        let mut context = EVM.entry(self.thread_name).or_default();
+        context.environment.clear_events();
+    }
+
+    /// Gets all emitted events for a test case.
+    pub fn events(self) -> Vec<Vec<u8>> {
+        let context = EVM.entry(self.thread_name).or_default();
+        context.environment.events()
+    }
+}
+
+#[derive(Default)]
+struct TestCase {
+    storage: MockStorage,
+    environment: Environment,
+}
+
+/// A global mutable key-value store mockig EVM behaviour.
+/// Allows concurrent access.
+///
+/// The key is the name of the test thread,
+/// and the value is the context of the test case.
+static EVM: Lazy<DashMap<ThreadName, TestCase>> = Lazy::new(DashMap::new);
```

Comment:
We can call it `MOCK_EVM`. But yeah, mentioning `EVM` in the name of the test storage seems quite reasonable to me!

Reply:
Since we are trying to mimic its behavior for test cases